### PR TITLE
add basic thread support

### PR DESCRIPTION
**Pull request by zhaozg**
Any advice, will with more feature,
come from uvbook, thread-create.lua
```lua
local uv = require('luv')

local hare_id = uv.new_thread()
local tortoise_id = uv.new_thread()
local step = 10

uv.thread_create(hare_id,function(step) 
    local ffi = require'ffi'

    local sleep 
    if ffi.os=='Windows' then
        ffi.cdef "void Sleep(int ms);"
        sleep = ffi.C.Sleep
    else 
        ffi.cdef "unsigned int usleep(unsigned int seconds);"
        sleep = ffi.C.usleep
    end
    while (step>0) do 
        step = step - 1
        sleep(1000)
        print("Hare ran another step")
    end
    print("Hare done running!")
end, step,true,'abcd','false')

uv.thread_create(tortoise_id,function(step)
    while (step>0) do 
        step = step - 0.5
        print("Tortoise ran another step")
    end
    print("Tortoise done running!")
end,step,'abcd','false')

print(hare_id==hare_id,uv.thread_equal(hare_id,hare_id))
print(tortoise_id==hare_id,uv.thread_equal(tortoise_id,hare_id))

uv.thread_join(hare_id)
uv.thread_join(tortoise_id)
```

### CODE REVIEW

**Comment by creationix** [comment]
Thanks for getting this started.  I'll look into it.

**Comment by creationix** [comment]
I'm a little concerned about threads having access to side-effects like `print`, especially if that's via libuv.

I think we could make something like web-workers quite safely.  We could even use `uv_async` to send messages between threads.

File: src/luv.c [hunk 1]
```diff
@@ -242,6 +242,13 @@ static const luaL_Reg luv_functions[] = {
   {"version", luv_version},
   {"version_string", luv_version_string},
 
+  // thread.c
+  {"new_thread", luv_new_thread},
+  {"thread_create", luv_thread_create}, 
+  {"thread_equal", luv_thread_equal},
+  {"thread_self", luv_thread_self},
+  {"thread_join", luv_thread_join},
+
```
Comment (creationix) [inline src/luv.c]: We should also add the method form for `:create`, `:equal`, and `:join` in the metatable for threads so you can do:

```lua
uv.new_thread():create(function () 
  -- ...
end)
```

Also, is there a reason to have `uv.thread_create` be apart from `uv.new_thread`?

**Comment by zhaozg** [comment]
Yes,thread and work need more think and good design,it it long term work,now just begin.

File: src/luv.c (hunk 1 above)
Comment (zhaozg) [inline src/luv.c]: New only create handle,and create real start thread,

Comment (zhaozg) [inline src/luv.c]: New only create handle,and create real start thread,please don't cost more time on it,because will  change lot in afew days.

**Comment by zhaozg** [comment]
It hard work to make lua script  as work_cb and after_work_cb. serial callback code everytime will a lots of CPU time

**Comment by zhaozg** [comment]
Hi @creationix,  When I read http://nikhilm.github.io/uvbook/threads.html#inter-thread-communication, I realize that 
uv_async_t is design for ITC(inter-thread-communication) , main loop thread create it and share it with thread or threadpool.
But according https://github.com/luvit/luv/blob/master/src/async.c#L31-L52 and https://github.com/luvit/luv/blob/master/docs.md#uv_async_t--async-handle
, after uv.new_async create async object,we have no method to pass to sub or others threads.

When try to code http://nikhilm.github.io/uvbook/threads.html#inter-thread-communication with luv, it cause async.c design or find a way share userdata.

In my options, async should behind in thread.c and work.c, maybe will broken freezed code.

**Comment by zhaozg** [comment]
work demo
```lua
local uv = require('luv')

local ctx = uv.new_work(
    function(n) --work,in threadpool
        local uv = require('luv')
        local t = uv.thread_self()
        print(t)
        uv.sleep(100)
        return n*n,n 
    end, 
    function(r,n) print(string.format('%d => %d',n,r)) end    --after work, in loop thread
)
uv.queue_work(ctx,2)
uv.queue_work(ctx,4)
uv.queue_work(ctx,4)
uv.queue_work(ctx,4)
uv.queue_work(ctx,4)

uv.run('default')
uv.loop_close()
```

**Pull request closed without merging**